syntax = "proto3";

enum TrainType {

    TRAIN_TYPE_UNSPECIFIED = 0;
    TRAIN_TYPE_HIGH_SPEED = 1;
    TRAIN_TYPE_REGULAR = 2;
}

// Company operating the train
enum Operator {

    OPERATOR_UNSPECIFIED = 0;
    OPERATOR_SOUTH_WESTERN_RAILWAY = 1;
    OPERATOR_THAMESLINK = 2;
    OPERATOR_EAST_COAST_TRAINS = 3;
  }

message Train {

  // ID is the train's unique identifier
  int32 id = 1;

  // Name of the train
  string name = 2;

  // The type of the train
  TrainType train_type = 3;

  // Company operating the train
  Operator operator = 4;

  // Number of coaches
  int32 coach_count = 5;

  // Whether the train has food facilities
  bool catering = 6;
}

// A service handles train management
service TrainAPI {

  rpc ListTrains(ListTrainsRequest) returns (ListTrainsResponse);
  rpc CreateTrain(CreateTrainRequest) returns (CreateTrainResponse);
  rpc DeleteTrain(DeleteTrainRequest) returns (DeleteTrainResponse);
}

// Create a new train
message CreateTrainRequest {

  // Train to create.
  Train train = 1;
}

message CreateTrainResponse {}

// Delete a train by id
message DeleteTrainRequest {

  // ID of train to be deleted
  int32 id = 1;
}

// Request a list of trains
message ListTrainsRequest {}

// Response containing a list of Trains
message ListTrainsResponse {

  // Trains to be returned
  repeated Train trains = 1;
}

message DeleteTrainResponse {}
